crates/network-monitor/src/lib.rs

```diff
@@ -0,0 +1,241 @@
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::collections::VecDeque;
+use std::fs;
+use std::future::Future;
+use std::io;
+use std::mem;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::time::Duration;
+
+use camino::*;
+use chrono::{DateTime, TimeDelta, Utc};
+use serde::*;
+use thiserror::Error;
+use tokio::select;
+use tokio_util::sync::CancellationToken;
+
+use network_scanner::ping;
+
+mod log_queue;
+mod state;
+
+pub use crate::state::State;
+
+#[derive(Error, Debug)]
+#[error(transparent)]
+pub enum SetConfigError {
+    Io(#[from] io::Error),
+    Serde(#[from] serde_json::Error),
+}
+
+pub async fn set_config(config: MonitorsConfig, state: Arc<State>) -> Result<(), SetConfigError> {
+    let file = fs::OpenOptions::new()
+        .create(true)
+        .write(true)
+        .truncate(true)
+        .open(get_monitor_config_path(&state.cache_path))?;
+
+    let mut config_write = state.config.write().await;
+
+    serde_json::to_writer_pretty(&file, &config)?;
+
+    let old_config = mem::replace(&mut *config_write, config);
+
+    let new_config_set: HashSet<&MonitorDefinition> = config_write.monitors.iter().collect();
+    let old_config_set: HashSet<&MonitorDefinition> = old_config.monitors.iter().collect();
+
+    let added = new_config_set.difference(&old_config_set);
+    let deleted = old_config_set.difference(&new_config_set);
+
+    let (new_cancellation_tokens, new_monitors): (
+        Vec<(String, CancellationToken)>,
+        Vec<Pin<Box<dyn Future<Output = ()> + Send>>>,
+    ) = added
+        .map(|definition| {
+            let cancellation_token = CancellationToken::new();
+            let cancellation_monitor = cancellation_token.clone();
+
+            let definition_clone = (*definition).clone(); // TODO: is there a nicer way to do this?
+
+            let state = state.clone();
+
+            let monitor = async move {
+                loop {
+                    let start_time = Utc::now();
+
+                    let monitor_result = match definition_clone.probe {
+                        ProbeType::Ping => do_ping_monitor(&definition_clone).await,
+                        ProbeType::TcpOpen => do_tcpopen_monitor(&definition_clone).await,
+                        ProbeType::Unknown(_) => return, // TODO: shouldn't happen, they should be filtered out. Create a separate ProbeType enum without Unknown?
+                    };
```

Comment:
When parsing the config, I have an `Unknown` case in the enum for the sake of forward-compatibility. I'd like to filter these out at the start of this function, in order to avoid dealing with the Unknown case here. Can I do that without having to manually write a duplicate of the `ProbeType` enum? Can I start with an unknown with no unknown case, and then derive a version that has it?

Reply:
> It will happen as we introduce new monitor types. In the long-term I'd like gateway to return some json-structured warning the client that some if its monitor types weren't supported, so I do want them to be represented in the array of monitors.

By the way, the preflight API does that.

Reply:
I've went ahead and separated the internal structs from the DTO, and now my issues are resolved. I also added a cookbook section, check it out!